core/providertracker/providertracker.go

```diff
@@ -0,0 +1,72 @@
+// Copyright 2024 Canonical Ltd.
+// Licensed under the AGPLv3, see LICENCE file for details.
+
+package providertracker
+
+import (
+	"context"
+
+	"github.com/juju/errors"
+
+	"github.com/juju/juju/environs"
+)
+
+// Provider in the intersection of a environs.Environ and a caas.Broker.
+//
+// We ideally don't want to import the environs package here, but I've not
```

Comment:
Worth adding comment detail to the effect of not being useful in and of itself, but serves as the constraint for generic parameterisation, preferable to `any`.